src/waiter.c

```diff
@@ -0,0 +1,311 @@
+                waiter->lock = srw_lock_create("waiter", false);
+                if (waiter->lock == NULL)
+                {
+                    LogError("Failure in srw_lock_create(\"waiter\", false)");
+                    result = NULL;
+                }
+                else
+                {
+                    THANDLE_INITIALIZE(RC_PTR)(&waiter->data, NULL);
+                    waiter->notification_callback = NULL;
+                    waiter->notification_callback_context = NULL;
+                    waiter->notify_complete_callback = NULL;
+                    waiter->notify_complete_callback_context = NULL;
+                    waiter->registered = false;
+                    waiter->notified = false;
+                    waiter->callbacks_called = false;
+                    result = waiter;
+                    goto all_ok;
+                }
+            }
+            sm_destroy(waiter->sm);
+        }
+        free(waiter);
+    }
+all_ok:
+    return result;
+}
+
+static void abandon(void* context)
+{
+    WAITER_HANDLE waiter = (WAITER_HANDLE)context;
+    srw_lock_acquire_exclusive(waiter->lock);
+    if (!waiter->callbacks_called)
+    {
+        sm_fault(waiter->sm);
+        if (waiter->registered)
+        {
+            waiter->notification_callback(waiter->notification_callback_context, NULL, WAITER_RESULT_ABANDONED);
+            sm_exec_end(waiter->sm);
+        }
+        if (waiter->notified)
+        {
+            waiter->notify_complete_callback(waiter->notify_complete_callback_context, WAITER_RESULT_ABANDONED);
+            THANDLE_ASSIGN(RC_PTR)(&waiter->data, NULL);
+            sm_exec_end(waiter->sm);
+        }
+        waiter->callbacks_called = true;
+    }
+    srw_lock_release_exclusive(waiter->lock);
+}
+
+void waiter_destroy(WAITER_HANDLE waiter)
+{
+    if (waiter == NULL)
+    {
+        LogError("Invalid argument: WAITER_HANDLE waiter=%p", waiter);
+    }
+    else
+    {
+        sm_close_begin_with_cb(waiter->sm, abandon, waiter, NULL, NULL);
+        srw_lock_destroy(waiter->lock);
+        sm_destroy(waiter->sm);
+        free(waiter);
+    }
+}
+
+static void cancel_register_notification(void* context)
+{
+    WAITER_HANDLE waiter = context;
+    srw_lock_acquire_exclusive(waiter->lock);
+    {
+        waiter->notification_callback(waiter->notification_callback_context, NULL, WAITER_RESULT_CANCELLED);
+        waiter->callbacks_called = true;
+        sm_fault(waiter->sm);
+        sm_exec_end(waiter->sm);
+    }
+    srw_lock_release_exclusive(waiter->lock);
+}
+
+static void dispose(void* context)
+{
+    (void)context;
+}
+
+static void try_complete(WAITER_HANDLE waiter)
+{
+    if (waiter->registered && waiter->notified && !waiter->callbacks_called)
+    {
+        sm_fault(waiter->sm);
+        waiter->callbacks_called = true;
+        sm_exec_end(waiter->sm);
+        waiter->notification_callback(waiter->notification_callback_context, waiter->data, WAITER_RESULT_OK);
```

Comment:
>notification_callback [](http://example.com/codeflow?start=16&length=21)

can this callback call any of the APIs of the module?

Reply:
No. No reentry allowed.